### ISSUE TITLE
High cpu if polling is being used in multiple devices

### ISSUE THREAD

**Issue by gui28347** [ztalbot2000/homebridge-cmd4#72]
Hi, 
   if you create 10 devices (contactSwitches for example), and use polling with 5 or 10 seconds it will cause 100% cpu, because the get state needs to run node to be able to read the txt files in fs.

   Would be possible to move the get to don't use one additional instance of node ? Such as integrating the Get command directly in main JS or do you have any other recommendation to avoid the high cpu ? Maybe adding an option to config like (dontusenodeforGet: yes/no) would fix this issue. 

I already tested the same scenario using the Get integrated into index.js and no issues, CPU 0%

btw: congrats for your solution it is really good for users using homebridge.

Thanks

**Comment by ztalbot2000** [comment]
Hi,

  Interesting delema. Polling is only if you want to have HomeKit to be
always up to date with the switches state; Otherwise the apps query the
device when you open them. So, polling could be removed or adjusted to a
larger interval, decreasing CPU.

Unfortunately, HomeKit is based an a single characteristic get/Set
methodology.  No other plugin does what you suggest, though it is not a bad
idea.  Let me sleep on it.

TTFN,
John

**Comment by gui28347** [comment]
Yep, mainly if you open two iphones at same time, (or the web interface) it start opening multiple instances of node to execute the getstates, and even with timeout of 30 seconds, multiple instances got timeout error message. I'm only with 10 devices, think about if you have 50 devices  your linux which is kind of crazy (# of devices X polling interval X #of devices using HomeKit), and number of reads ("getState") will be always higher than setStates anyway... so adding an if statement to your driver makes it even better, because the idea of keeping the txt files as cache is amazing, when the system is rebooted or resume the plugin can recover last status using the txt files.  

So in summary the suggestion is:

  Having a parameter in config:   dontusenodeforGet (true/false)

   If parameter is set to false, exactly same code you have today. (no code changes)

   If parameter is set to true is just execute readData function you have in State.js (migrate it to index.js) . This way the performance will be much better. Even when not using 10 devices, every time when I open the phone I saw the devices "Updating" , now using the getState without creating more instances of node it is super fast, I don't see the message "Updating" in devices... 

  I think this can help other users using your amazing plugin :-) 
Thanks
Gui

**Comment by ztalbot2000** [comment]
Hi,

  I did think about your predicament last night.  I will have to think
again about your new idea.  What I did think of last night was changing
state_cmd to something like a batch_cmd, so that one get could return
multiple characteristics when polling. For example:
Get <Device> <characteristic> <characteristic> ...
or maybe:
Get <Device> AllCharacteristics
Cmd4 would then have to parse the  One response for each value, sending
that to Homebridge as if multiple Gets were done.


I'm still thinking,

TTFN,
John

**Comment by gui28347** [comment]
Yep, doing in batches will help a lot because it will reduce # of child processes running.

But do you see any benefit of creating the child processes? If in your index.js you can simple read  without creating the child process, what is the benefit of keep the State.js being called rather than directly read files ?

I believe for performance reasons if you can avoid creating child process at all will be better. In some scenarios I know that is not possible...  

Is the state.js doing any other action, rather than just reading values in text files and sending it back to caller ?

**Comment by ztalbot2000** [comment]
Hi,

    You are correct. It would work in your case. If the state was always
available from a file, the file could just be read.  That would mean that
you as a user of CMD4 would have to leave the state in a file for cmd4 to
read.  There would still have to be a process that you would create to have
the state put in the file. The difference is that having CMD4 do the query
would mean that the state is what it is when CMD4 asks for it. Wherever the
process is, it is still a process that has to run.
   Interesting thought though.  It could certainly be easily done.  Another
option perhaps.  Do you want to spin off your own version?

ttfn,
John

**Comment by gui28347** [comment]
I was thinking about it. Since this change will also benefit other users that want to create multiple sensors
I think would be great if you can add to your code, mainly if you will apply for home bridge verified, it will bring more visibility to your plugin. 

No rush on changing it,  just add to it to your backlog of feature requests.

**Comment by ztalbot2000** [comment]
Hi,

   Homebridge verified requires a config.schéma that would take months to
create for this plugin.  It will not happen any time soon.  There is
already a feature request for this. Sorry.

TTFN,
John

**Comment by gui28347** [comment]
After several tests, I found the best performance scenario without using files at all.

1) Use a in memory dictionary to save the values. This way you have have ton of accessories and even with 1 second polling in all of them, it will be using memory only.

2) To be able to interact with in memory dictionary I believe using an HTTP listening port to receive commands from Cmd4 Scripts to read/write variables will be the best.

Using this approach this solution scale to hundreds of accessories and don't have the limitations of using polling , causing either high CPU (because the multiple nodejs instances) or potential high IO if read/write functions are integrated in the main code.

Using dictionary it will always keep last status in memory, so even using 1 second refresh will cause zero issues.

If you like the idea, I can also help implement it.

**Comment by ztalbot2000** [comment]
Hi,

  Sorry for not getting back to you.  My health is not the greatest, but I
have been working hard on Cmd4.  I have a dev branch that I have a 3.0
version in progress.  Lots of stuff there to implement CMD4 as Homebridge
suggests. It is close to release, except that restarts are an issue because
I have yet to get the code working  to  re-read saved config changes ( like
name for instance ) and I only did a first pass on the documentation.

I do like your idea and have something in mind along the same lines.  Since
Cmd4 is highly in flux, please wait before attempting your own.  I can give
you a heads up when b_IOS14_dev branch is ready.

TTFN,
John

**Comment by ztalbot2000** [comment]
Hi,

I just submitted into dev the solution.  Actually there are three
possibilities.
1. "fetch": "Always",    - As before
2. "fetch": "Cached", - Only fetch the cached valus.  Polling would have to
update those values.
3. "fetch": "Polled", - All characters are fetched from Cache, except those
that are set for Polling, which are done as they were before.

I like this solution because it's backward compatible and I don't muck with
getValue at runtime. During init a new bind for "get" from HomeKit is
created to retrieve the cached value directly.

This 3.0 update in my b_IOS14Dev branch has a lot of stuff. It all passes
unit testing and some limited regression testing. Hopefully it will be
available soon.

TTFN,
John Talbot

**Comment by ztalbot2000** [comment]
Hi,

 I just merged Cmd4 Version 3.0.0 into GitHub, putting that horrendous label on it too.
 it solves the TV icon with the caveate that you add the option of: publishExternally: true as well as the high CPU occupancy with the fetch options, I previously mentioned and now have fully documented.  If you read the documentation, you will understand why I decided on the label..
 In many ways it does not deserve the label because the state_cmd does not change and neither do the libs.  Cmd4, looks almost exactly like the Homebridge template plugin they provide and with additional Unit testing and now linted as well, I hope you give it a try as I soak it a few more days before submitting it to npm.

 thanks,
 John Talbot

**Comment by ztalbot2000** [comment]
Just a note on this one. I shall be releasing 3.3 soon. Polling will call updateValue as it is supposed to for background changes meaning polling will be totally indépendant from fetch.  You will then be able to have everything fetch cached and only have polling update the values, if you so wish.

take care,
John Talbot
